File: yandex/coderun/frontend/knight-move/knight-move.py
import sys
from typing import List

def main():
    n_row, n_col = input().split()
    n_row = int(n_row)
    n_col = int(n_col)
    mem = [[-1] * n_col for _ in range(n_row)]
    mem[n_row-1][n_col-1] = 1
    print(solve(n_row, n_col, 0, 0, mem))


def solve(n_row: int, n_col: int, i: int, j: int, mem: List[List[int]]) -> int:
    if 0 <= i < n_row and 0 <= j < n_col:
        if mem[i][j] < 0:
            ans = 0
            if i < n_row-2 and j < n_col-1:
                ans += solve(n_row, n_col, i+2, j+1, mem)
            if i < n_row-1 and j < n_col-2:
                ans += solve(n_row, n_col, i+1, j+2, mem)
            mem[i][j] = ans
        return mem[i][j]
    else:
        return 0



if __name__ == '__main__':
  main()

File: yandex/coderun/frontend/knight-move/knight-move.js
// Для чтения входных данных в Node.js необходимо использовать
// модуль readline, который работает с потоком ввода-вывода
// (stdin/stdout) и позволяет читать строки.
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Данные во входном потоке могут состоять из нескольких строк.
// Чтобы прочитать их, можно использовать метод rl.on(),
// который вызывается каждый раз при появлении новой строки
// в потоке ввода.
// Чтобы вывести результат в поток вывода (stdout),
// можно использовать метод console.log().
// Пример:
// console.log('Результат:', result);

rl.on('line', line => {
  function solve(n, m, i, j, mem) {
    if (mem[i][j] < 0){
      let ans = 0;
      if (i < n-2 && j < m-1) {
        ans += solve(n, m, i+2, j+1, mem);
      }
      if (i < n-1 && j < m-2) {
        ans += solve(n,m, i+1, j+2, mem);
      }
      mem[i][j] = ans;
    }
    return mem[i][j]
  }
  const [N, M] = line.split(' ').map(Number);
  let mem = new Array(N);            // создаем пустой массив длины `M`
  for (let i = 0; i < N; i++) {
    mem[i] = new Array(M).fill(-1);        // делаем каждый элемент массивом
  }
  mem[N-1][M-1] = 1;


  solve(N, M, 0, 0, mem);
  console.log(solve(N, M, 0, 0, mem));

 rl.close();
});


